File: py/src/domain/value_objects/currency.py
class Currency:
    def __init__(self, code: str):
        clean_code = code.strip().upper()
        if len(clean_code) != 3:
            raise ValueError("Currency code must be a 3-letter ISO code.")
        
        supported = {"USD", "EUR", "INR"}
        if clean_code not in supported:
            raise ValueError(f"Unsupported currency code: {clean_code}. Supported: {', '.join(supported)}")
        
        self._code = clean_code

    @property
    def code(self) -> str:
        return self._code

    def equals(self, other: 'Currency') -> bool:
        if not isinstance(other, Currency):
            return False
        return self._code == other.code

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Currency):
            return NotImplemented
        return self.equals(other)

    def __repr__(self) -> str:
        return f"Currency({self._code})"

File: js/src/domain/value-objects/Currency.ts
export class Currency {
  private constructor(public readonly code: string) {}

  public static fromCode(code: string): Currency {
    const cleanCode = code.trim().toUpperCase();
    if (cleanCode.length !== 3) {
      throw new Error("Currency code must be a 3-letter ISO code.");
    }
    // For our system, we support USD, EUR, and INR
    const supported = ["USD", "EUR", "INR"];
    if (!supported.includes(cleanCode)) {
      throw new Error(`Unsupported currency code: ${cleanCode}. Supported: ${supported.join(", ")}`);
    }
    return new Currency(cleanCode);
  }

  public equals(other: Currency): boolean {
    return this.code === other.code;
  }
}
